### PR TITLE
feat: Poseidon2-based Sponge function

### PR DESCRIPTION
**Pull request by alxiong**
Part of: #709 (Item 1 and 3)

### This PR: 

- Implement a generic P2-based Sponge struct which implements `nimue`'s `trait Sponge`
- Concretely instantiate `DuplexSponge` for all the parameters we have (including various rate, given the same state size)
- Add tests for all (both functional correctness as sponge used in transcript, and a naive statistical tests that output of these sponge-squeezed bytes are close to uniform distribution)

### Caveats:

`nimue-v0.1.0` which is the only published version, has compilation error, which I have reported upstream and help submit a PR to fix it. 
I will push for a release tag of `0.1.1` from the upstream and later repin our `nimue` dependency, for now, I'm pointing to my own fork/branch

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (main)
- [x] Linked to GitHub issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Wrote unit tests
- [x] Updated relevant documentation in the code
- [ ] Added relevant changelog entries to the `CHANGELOG.md` of touched crates.
- [ ] Re-reviewed `Files changed` in the GitHub PR explorer

### CODE REVIEW

**Comment by alxiong** [comment]
I have no clue about the CI error, firstly we didn't use getrandom 0.2.15, (in my local cargo.lock, we only depends on 0.2.12); secondly, that error shouldn't come from any of my changes. 

We can safely review the code and ignore this error for now. (I will figure it out later)

File: merkle_tree/Cargo.toml [hunk 1]
```diff
@@ -24,7 +24,7 @@ displaydoc = { workspace = true }
 hashbrown = { workspace = true }
 hex = "0.4.3"
 itertools = { workspace = true, features = ["use_alloc"] }
-jf-poseidon2 = { path = "../poseidon2" }
+jf-poseidon2 = { version = "0.1.0", git = "https://github.com/EspressoSystems/jellyfish" }
```
Comment (mrain) [inline merkle_tree/Cargo.toml:27]: should we also have a version number bump for jf-merkle-tree after CRHF is done?

**Review by mrain — APPROVED** [review]
LGTM!

File: merkle_tree/Cargo.toml (hunk 1 above)
Comment (alxiong) [reply merkle_tree/Cargo.toml:27]: yes, I'll introduce the CRHF in the next PR, then another one after that to bump jf-merkle-tree

**Review by mrain — APPROVED** [review]

**Pull request merged**